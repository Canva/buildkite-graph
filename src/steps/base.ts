import ow from 'ow';
import 'reflect-metadata';
import { Expose, Exclude, Transform } from 'class-transformer';

export interface BaseStep {}

// see https://github.com/microsoft/TypeScript/issues/22815#issuecomment-375766197
export interface DefaultStep extends BaseStep {}
export abstract class DefaultStep implements BaseStep {
    @Exclude()
    public readonly dependencies: Set<DefaultStep> = new Set();

    dependsOn(step: DefaultStep): this {
        this.dependencies.add(step);
        return this;
    }

    @Exclude()
    public always: boolean = false;

    alwaysExecute() {
        this.always = true;
        return this;
    }
}

@Exclude()
export class BranchLimitedStep extends DefaultStep {
    @Expose({ name: 'branches' })
    @Transform((branches: Set<string>) =>
        branches.size ? [...branches].sort().join(' ') : undefined,
    )
    private _branches: Set<string> = new Set();

    withBranch(pattern: string): this {
        ow(pattern, ow.string.nonEmpty);
        this._branches.add(pattern);
        return this;
    }
}

export class LabeledStep extends BranchLimitedStep {
    private _label?: string;

    @Expose()
    get label() {
        return this._label;
    }

    withLabel(label: string) {
        this._label = label;
        return this;
    }
}

@Exclude()
export abstract class Chainable<T> {
    constructor(protected readonly parent: T) {
        this.parent = parent;
    }
}

export type ExitStatus = number | '*';

export const exitStatusPredicate = ow.any(
    ow.string.equals('*'),
    ow.number.integer,
);
