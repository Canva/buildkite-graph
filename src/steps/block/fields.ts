import { Exclude, Expose } from 'class-transformer';
import ow from 'ow';
import { Chainable } from '../../base';

@Expose()
abstract class Field {
    public readonly required?: false;
    constructor(
        public readonly key: string,
        public readonly hint?: string,
        required = false,
    ) {
        ow(key, ow.string.nonEmpty);
        ow(key, ow.string.matches(/[0-9a-z-\/]+/i));
        if (!required) {
            this.required = required;
        }
    }
}
@Expose()
export class TextField extends Field {
    private readonly text: string;
    private readonly default?: string;
    constructor(
        key: string,
        label: string,
        hint?: string,
        required = true,
        defaultValue?: string,
    ) {
        super(key, hint, required);
        this.text = label;
        this.default = defaultValue;
    }
}
export class Option {
    constructor(
        private readonly label: string,
        private readonly value: string,
    ) {
        ow(label, ow.string.nonEmpty);
        ow(value, ow.string.nonEmpty);
    }
}
@Expose()
export class SelectField extends Field {
    private readonly select: string;
    private options: Option[] = [];
    private readonly multiple?: true;
    private readonly default?: string | string[];
    constructor(
        key: string,
        label: string,
        hint?: string,
        required?: boolean,
        multiple?: false,
        defaultValue?: string,
    );
    constructor(
        key: string,
        label: string,
        hint?: string,
        required?: boolean,
        multiple?: true,
        defaultValue?: string | string[],
    );
    constructor(
        key: string,
        label: string,
        hint?: string,
        required = true,
        multiple = false,
        defaultValue?: string | string[],
    ) {
        super(key, hint, required);
        this.select = label;
        this.default = defaultValue;
        if (multiple) {
            this.multiple = multiple;
        }
    }
    addOption(option: Option): this {
        this.options.push(option);
        return this;
    }
}
export interface Fields<T> {
    add(field: Field): T;
}

@Exclude()
export class FieldsImpl<T> extends Chainable<T> implements Fields<T> {
    fields: Map<string, Field> = new Map();
    add(field: Field): T {
        this.fields.set(field.key, field);
        return this.parent;
    }
    hasFields(): boolean {
        return this.fields.size > 0;
    }
}
