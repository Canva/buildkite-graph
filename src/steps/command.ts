import { Exclude, Expose, Transform } from 'class-transformer';
import ow from 'ow';
import 'reflect-metadata';
import { KeyValue, KeyValueImpl, transformKeyValueImpl } from '../key_value';
import {
    Plugin,
    Plugins,
    PluginsImpl,
    transformPlugins,
} from './command/plugins';
import { ExitStatus, exitStatusPredicate, LabeledStep } from '../base';
import { Retry, RetryImpl } from './command/retry';

function assertTimeout(timeout: number): void {
    ow(timeout, ow.number.integerOrInfinite.positive);
}

function assertSkipValue(value: SkipValue): void {
    if (typeof value === 'string') {
        ow(value, ow.string.nonEmpty);
    }
}

type SkipValue = boolean | string;
export type SkipFunction = () => SkipValue;

export class Command {
    constructor(public command: string, public timeout: number = Infinity) {
        ow(command, ow.string);
        assertTimeout(timeout);
    }

    toString(): string {
        return this.command;
    }
}

type Agents = Map<string, string>;

@Exclude()
export class CommandStep extends LabeledStep {
    @Expose({ name: 'command' })
    @Transform((value: Command[]) => {
        if (!value || value.length === 0) {
            return undefined;
        }
        return value.length === 1
            ? value[0].command
            : value.map(c => c.command);
    })
    public readonly command: Command[] = [];

    @Expose()
    @Transform(transformKeyValueImpl)
    public readonly env: KeyValue<this>;

    private _id?: string;
    @Expose({ name: 'id' })
    private get id(): string | undefined {
        return this._id;
    }

    private _parallelism?: number;

    @Expose({ name: 'parallelism' })
    private get paralellism(): number | undefined {
        return this._parallelism;
    }

    @Expose()
    private concurrency?: number;

    @Expose({ name: 'concurrency_group' })
    private concurrencyGroup?: string;

    @Expose({ name: 'artifact_paths' })
    @Transform((paths: Set<string>) => (paths.size ? paths : undefined))
    private _artifactPaths: Set<string> = new Set();

    private _agents: Agents = new Map();

    @Expose()
    @Transform((agents: Agents) => (agents.size ? agents : undefined))
    get agents(): Agents {
        return this._agents;
    }

    private _timeout?: number;

    @Expose({ name: 'timeout_in_minutes' })
    get timeout(): number | undefined {
        if (this._timeout === Infinity || this._timeout === 0) {
            return undefined;
        } else if (this._timeout) {
            return this._timeout;
        }
        if (this.command) {
            const value = this.command.reduce((acc, command) => {
                acc += command.timeout;
                return acc;
            }, 0);
            if (value === Infinity || value === 0) {
                return undefined;
            } else if (value) {
                return value;
            }
        }
    }

    @Expose({ name: 'plugins' })
    @Transform(transformPlugins)
    public readonly plugins: Plugins<this> = new PluginsImpl(this);

    @Expose({ name: 'soft_fail' })
    @Transform((value: Set<ExitStatus>) => {
        if (!value.size) {
            return undefined;
        } else if (value.has('*')) {
            return true;
        } else {
            return [...value].map(s => ({
                // eslint-disable-next-line @typescript-eslint/camelcase
                exit_status: s,
            }));
        }
    })
    private _softFail: Set<ExitStatus> = new Set();

    @Expose({ name: 'skip' })
    @Transform((value: SkipValue | SkipFunction) => {
        if (typeof value === 'function') {
            value = value();
            assertSkipValue(value);
        }
        return value || undefined;
    })
    private _skip?: SkipValue | SkipFunction;

    @Expose()
    @Transform((value: RetryImpl<any>) =>
        value.hasValue() ? value : undefined,
    )
    public readonly retry: Retry<CommandStep> = new RetryImpl(this);

    constructor(plugin: Plugin, label?: string);
    constructor(command: string, label?: string);
    constructor(command: Command, label?: string);
    constructor(commands: (string | Plugin | Command)[], label?: string);
    constructor(
        command: string | Plugin | Command | (string | Plugin | Command)[],
        label?: string,
    ) {
        super();
        this.env = new KeyValueImpl(this);
        if (label) {
            this.withLabel(label);
        }
        if (Array.isArray(command)) {
            ow(command, ow.array.minLength(1));
            for (const c of command) {
                this.add(c);
            }
        } else {
            this.add(command);
        }
    }

    add(c: string | Plugin | Command): this {
        if (c instanceof Plugin) {
            this.plugins.add(c);
        } else if (typeof c === 'string') {
            this.command.push(new Command(c));
        } else if (c instanceof Command) {
            this.command.push(c);
        }
        return this;
    }

    /**
     *
     * @param timeout for the step in seconds.
     */
    withTimeout(timeout = Infinity): this {
        assertTimeout(timeout);
        this._timeout = timeout;
        return this;
    }

    withParallelism(parallelism: number): this {
        ow(parallelism, ow.number.integer.positive);
        this._parallelism = parallelism;
        return this;
    }

    withAgent(key: string, value: string): this {
        ow(key, ow.string.nonEmpty);
        ow(value, ow.string.nonEmpty);
        this._agents.set(key, value);
        return this;
    }

    withArtifactPath(...globs: string[]): this {
        ow(globs, ow.array.ofType(ow.string.nonEmpty));
        for (const glob of globs) {
            this._artifactPaths.add(glob);
        }
        return this;
    }

    withConcurrency(concurrency: number, group: string): this {
        ow(concurrency, ow.number.integer.positive);
        this.concurrency = concurrency;
        this.concurrencyGroup = group;
        return this;
    }

    withId(identifier: string): this {
        ow(identifier, ow.string.nonEmpty);
        this._id = identifier;
        return this;
    }

    withSoftFail(fail: ExitStatus | true): this {
        if (fail !== true) {
            ow(fail, exitStatusPredicate);
        }
        this._softFail.add(fail === true ? '*' : fail);
        return this;
    }

    skip(skip: SkipValue | SkipFunction): this {
        if (typeof skip !== 'function') {
            assertSkipValue(skip);
        }

        this._skip = skip;
        return this;
    }

    toString(): string {
        return (
            this.label ||
            (this.command
                ? `<${this.command.join(' && ') || '(empty)'}>`
                : this.plugins.toString())
        );
    }
}
