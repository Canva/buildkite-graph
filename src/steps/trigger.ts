import { Pipeline, Serializable } from '../index';
import { LabeledStep } from '../base';
import { Build, BuildImpl } from './trigger/build';

export class TriggerStep extends LabeledStep implements Serializable {
    get trigger(): string {
        return this._trigger instanceof Pipeline
            ? this._trigger.slug()
            : this._trigger;
    }

    private _async = false;

    public readonly build: Build<TriggerStep> = new BuildImpl(this);

    constructor(
        private readonly _trigger: Pipeline | string,
        label?: string,
        async = false,
    ) {
        super();
        if (label) {
            this.withLabel(label);
        }
        this._async = async;
    }
    async(async: boolean): this {
        this._async = async;
        return this;
    }
    toString(): string {
        return this.label || `[trigger ${this.trigger}]`;
    }

    async toJson(): Promise<object> {
        return {
            ...(await super.toJson()),
            trigger: this.trigger,
            async: this._async || undefined,
            build: await (this.build as BuildImpl<this>).toJson(),
        };
    }
}
