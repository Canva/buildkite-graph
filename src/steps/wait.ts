import { BaseStep } from '../base';
import { Serializable } from '../index';

export class WaitStep implements BaseStep, Serializable {
  constructor(public continueOnFailure = false) {}

  toString(): string {
    /* istanbul ignore next */
    return this.continueOnFailure ? '[wait; continues on failure]' : '[wait]';
  }

  async toJson(): Promise<Record<string, unknown>> {
    return {
      wait: null,
      continue_on_failure: this.continueOnFailure || undefined,
    };
  }
}
