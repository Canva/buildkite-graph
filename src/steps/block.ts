import { Exclude, Expose, Transform } from 'class-transformer';
import { BranchLimitedStep } from '../base';
import { Fields, FieldsImpl } from './block/fields';

@Exclude()
export class BlockStep extends BranchLimitedStep {
    @Expose({ name: 'block' })
    private readonly title: string;
    @Expose()
    private readonly prompt?: string;
    @Expose()
    @Transform((value: FieldsImpl<any>) =>
        value.hasFields() ? [...value.fields.values()] : undefined,
    )
    public readonly fields: Fields<this> = new FieldsImpl(this);
    constructor(title: string, prompt?: string) {
        super();
        this.title = title;
        this.prompt = prompt;
    }
    toString() {
        return `[block for '${this.title}']`;
    }
}
