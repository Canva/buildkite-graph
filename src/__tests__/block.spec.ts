import { Pipeline } from '../';
import { BlockStep } from '../steps/block';
import { TextField, SelectField, Option } from '../steps/block/fields';
import { createTest } from './helpers';

describe('buildkite-graph', () => {
    describe('Steps', () => {
        describe('Block', () => {
            createTest('can be added', () => [
                new Pipeline('whatever').add(
                    new BlockStep(':rocket: Release!'),
                ),
                new Pipeline('whatever').add(
                    new BlockStep(
                        ':rocket: Release!',
                        'Release to production?',
                    ),
                ),
            ]);

            createTest('branches', () =>
                new Pipeline('whatever').add(
                    new BlockStep('my title')
                        .withBranch('master')
                        .withBranch('stable/*')
                        .withBranch('!release/*'),
                ),
            );

            createTest('with fields', () => [
                new Pipeline('whatever').add(
                    new BlockStep('my title').fields
                        .add(new TextField('field-1'))
                        .fields.add(new SelectField('field-2'))
                        .fields.add(new TextField('field-3')),
                ),
                new Pipeline('whatever').add(
                    new BlockStep('my title').fields.add(
                        new TextField(
                            'release-name',
                            'Code Name',
                            'What’s the code name for this release? :name_badge:',
                            false,
                            'Flying Dolphin',
                        ),
                    ),
                ),
                new Pipeline('whatever').add(
                    new BlockStep('my title').fields.add(
                        new SelectField(
                            'release-stream',
                            'Stream',
                            'What’s the release stream?',
                            false,
                            false,
                            'beta',
                        )
                            .addOption(new Option('Beta', 'beta'))
                            .addOption(new Option('Stable', 'stable')),
                    ),
                ),
                new Pipeline('whatever').add(
                    new BlockStep('my title').fields.add(
                        new SelectField(
                            'regions',
                            'Regions',
                            'Which regions should we deploy this to? :earth_asia:',
                            true,
                            true,
                            ['na', 'eur', 'asia', 'aunz'],
                        )
                            .addOption(new Option('North America', 'na'))
                            .addOption(new Option('Europe', 'eur'))
                            .addOption(new Option('Asia', 'asia'))
                            .addOption(new Option('Oceania', 'aunz')),
                    ),
                ),
            ]);
        });
    });
});
