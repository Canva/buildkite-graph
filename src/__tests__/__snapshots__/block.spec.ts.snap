// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildkite-graph Steps Block branches dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for 'my title']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block branches json 1`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "branches": "!release/* master stable/*",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block branches json_depends_on 1`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "branches": "!release/* master stable/*",
      "key": "_0",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block branches json_depends_on_accept_all 1`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "branches": "!release/* master stable/*",
      "key": "_0",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block branches structure 1`] = `"* [block for 'my title']"`;

exports[`buildkite-graph Steps Block branches yaml 1`] = `
"steps:
  - branches: '!release/* master stable/*'
    block: my title
"
`;

exports[`buildkite-graph Steps Block branches yaml_depends_on 1`] = `
"steps:
  - branches: '!release/* master stable/*'
    key: _0
    block: my title
"
`;

exports[`buildkite-graph Steps Block branches yaml_depends_on_accept_all 1`] = `
"steps:
  - branches: '!release/* master stable/*'
    key: _0
    block: my title
"
`;

exports[`buildkite-graph Steps Block branches yaml_mutate 1`] = `
"steps:
  - branches: '!release/* master stable/*'
    key: _0
    block: my title
"
`;

exports[`buildkite-graph Steps Block can be added dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for ':rocket: Release!']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block can be added dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for ':rocket: Release!']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block can be added json 1`] = `
Object {
  "steps": Array [
    Object {
      "block": ":rocket: Release!",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block can be added json 2`] = `
Object {
  "steps": Array [
    Object {
      "block": ":rocket: Release!",
      "prompt": "Release to production?",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block can be added json_depends_on 1`] = `
Object {
  "steps": Array [
    Object {
      "block": ":rocket: Release!",
      "key": "_0",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block can be added json_depends_on 2`] = `
Object {
  "steps": Array [
    Object {
      "block": ":rocket: Release!",
      "key": "_1",
      "prompt": "Release to production?",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block can be added json_depends_on_accept_all 1`] = `
Object {
  "steps": Array [
    Object {
      "block": ":rocket: Release!",
      "key": "_0",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block can be added json_depends_on_accept_all 2`] = `
Object {
  "steps": Array [
    Object {
      "block": ":rocket: Release!",
      "key": "_1",
      "prompt": "Release to production?",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block can be added structure 1`] = `"* [block for ':rocket: Release!']"`;

exports[`buildkite-graph Steps Block can be added structure 2`] = `"* [block for ':rocket: Release!']"`;

exports[`buildkite-graph Steps Block can be added yaml 1`] = `
"steps:
  - block: ':rocket: Release!'
"
`;

exports[`buildkite-graph Steps Block can be added yaml 2`] = `
"steps:
  - block: ':rocket: Release!'
    prompt: Release to production?
"
`;

exports[`buildkite-graph Steps Block can be added yaml_depends_on 1`] = `
"steps:
  - key: _0
    block: ':rocket: Release!'
"
`;

exports[`buildkite-graph Steps Block can be added yaml_depends_on 2`] = `
"steps:
  - key: _1
    block: ':rocket: Release!'
    prompt: Release to production?
"
`;

exports[`buildkite-graph Steps Block can be added yaml_depends_on_accept_all 1`] = `
"steps:
  - key: _0
    block: ':rocket: Release!'
"
`;

exports[`buildkite-graph Steps Block can be added yaml_depends_on_accept_all 2`] = `
"steps:
  - key: _1
    block: ':rocket: Release!'
    prompt: Release to production?
"
`;

exports[`buildkite-graph Steps Block can be added yaml_mutate 1`] = `
"steps:
  - key: _0
    block: ':rocket: Release!'
"
`;

exports[`buildkite-graph Steps Block can be added yaml_mutate 2`] = `
"steps:
  - key: _1
    block: ':rocket: Release!'
    prompt: Release to production?
"
`;

exports[`buildkite-graph Steps Block with fields dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for 'my title']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block with fields dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for 'my title']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block with fields dot 3`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for 'my title']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block with fields dot 4`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[block for 'my title']\\" [ color = \\"grey\\" ];
}

}
"
`;

exports[`buildkite-graph Steps Block with fields json 1`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "key": "field-1",
          "text": "Label 1",
        },
        Object {
          "key": "field-2",
          "options": Array [],
          "select": "Label 2",
        },
        Object {
          "key": "field-3",
          "text": "Label 3",
        },
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json 2`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": "Flying Dolphin",
          "hint": "What’s the code name for this release? :name_badge:",
          "key": "release-name",
          "required": false,
          "text": "Code Name",
        },
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json 3`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": "beta",
          "hint": "What’s the release stream?",
          "key": "release-stream",
          "options": Array [
            Object {
              "label": "Beta",
              "value": "beta",
            },
            Object {
              "label": "Stable",
              "value": "stable",
            },
          ],
          "required": false,
          "select": "Stream",
        },
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json 4`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": Array [
            "na",
            "eur",
            "asia",
            "aunz",
          ],
          "hint": "Which regions should we deploy this to? :earth_asia:",
          "key": "regions",
          "multiple": true,
          "options": Array [
            Object {
              "label": "North America",
              "value": "na",
            },
            Object {
              "label": "Europe",
              "value": "eur",
            },
            Object {
              "label": "Asia",
              "value": "asia",
            },
            Object {
              "label": "Oceania",
              "value": "aunz",
            },
          ],
          "select": "Regions",
        },
      ],
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on 1`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "key": "field-1",
          "text": "Label 1",
        },
        Object {
          "key": "field-2",
          "options": Array [],
          "select": "Label 2",
        },
        Object {
          "key": "field-3",
          "text": "Label 3",
        },
      ],
      "key": "_0",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on 2`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": "Flying Dolphin",
          "hint": "What’s the code name for this release? :name_badge:",
          "key": "release-name",
          "required": false,
          "text": "Code Name",
        },
      ],
      "key": "_1",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on 3`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": "beta",
          "hint": "What’s the release stream?",
          "key": "release-stream",
          "options": Array [
            Object {
              "label": "Beta",
              "value": "beta",
            },
            Object {
              "label": "Stable",
              "value": "stable",
            },
          ],
          "required": false,
          "select": "Stream",
        },
      ],
      "key": "_2",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on 4`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": Array [
            "na",
            "eur",
            "asia",
            "aunz",
          ],
          "hint": "Which regions should we deploy this to? :earth_asia:",
          "key": "regions",
          "multiple": true,
          "options": Array [
            Object {
              "label": "North America",
              "value": "na",
            },
            Object {
              "label": "Europe",
              "value": "eur",
            },
            Object {
              "label": "Asia",
              "value": "asia",
            },
            Object {
              "label": "Oceania",
              "value": "aunz",
            },
          ],
          "select": "Regions",
        },
      ],
      "key": "_3",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on_accept_all 1`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "key": "field-1",
          "text": "Label 1",
        },
        Object {
          "key": "field-2",
          "options": Array [],
          "select": "Label 2",
        },
        Object {
          "key": "field-3",
          "text": "Label 3",
        },
      ],
      "key": "_0",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on_accept_all 2`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": "Flying Dolphin",
          "hint": "What’s the code name for this release? :name_badge:",
          "key": "release-name",
          "required": false,
          "text": "Code Name",
        },
      ],
      "key": "_1",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on_accept_all 3`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": "beta",
          "hint": "What’s the release stream?",
          "key": "release-stream",
          "options": Array [
            Object {
              "label": "Beta",
              "value": "beta",
            },
            Object {
              "label": "Stable",
              "value": "stable",
            },
          ],
          "required": false,
          "select": "Stream",
        },
      ],
      "key": "_2",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields json_depends_on_accept_all 4`] = `
Object {
  "steps": Array [
    Object {
      "block": "my title",
      "fields": Array [
        Object {
          "default": Array [
            "na",
            "eur",
            "asia",
            "aunz",
          ],
          "hint": "Which regions should we deploy this to? :earth_asia:",
          "key": "regions",
          "multiple": true,
          "options": Array [
            Object {
              "label": "North America",
              "value": "na",
            },
            Object {
              "label": "Europe",
              "value": "eur",
            },
            Object {
              "label": "Asia",
              "value": "asia",
            },
            Object {
              "label": "Oceania",
              "value": "aunz",
            },
          ],
          "select": "Regions",
        },
      ],
      "key": "_3",
    },
  ],
}
`;

exports[`buildkite-graph Steps Block with fields structure 1`] = `"* [block for 'my title']"`;

exports[`buildkite-graph Steps Block with fields structure 2`] = `"* [block for 'my title']"`;

exports[`buildkite-graph Steps Block with fields structure 3`] = `"* [block for 'my title']"`;

exports[`buildkite-graph Steps Block with fields structure 4`] = `"* [block for 'my title']"`;

exports[`buildkite-graph Steps Block with fields yaml 1`] = `
"steps:
  - block: my title
    fields:
      - key: field-1
        text: Label 1
      - key: field-2
        options: []
        select: Label 2
      - key: field-3
        text: Label 3
"
`;

exports[`buildkite-graph Steps Block with fields yaml 2`] = `
"steps:
  - block: my title
    fields:
      - key: release-name
        hint: 'What’s the code name for this release? :name_badge:'
        required: false
        text: Code Name
        default: Flying Dolphin
"
`;

exports[`buildkite-graph Steps Block with fields yaml 3`] = `
"steps:
  - block: my title
    fields:
      - key: release-stream
        hint: What’s the release stream?
        required: false
        options:
          - label: Beta
            value: beta
          - label: Stable
            value: stable
        select: Stream
        default: beta
"
`;

exports[`buildkite-graph Steps Block with fields yaml 4`] = `
"steps:
  - block: my title
    fields:
      - key: regions
        hint: 'Which regions should we deploy this to? :earth_asia:'
        options:
          - label: North America
            value: na
          - label: Europe
            value: eur
          - label: Asia
            value: asia
          - label: Oceania
            value: aunz
        select: Regions
        default:
          - na
          - eur
          - asia
          - aunz
        multiple: true
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on 1`] = `
"steps:
  - key: _0
    block: my title
    fields:
      - key: field-1
        text: Label 1
      - key: field-2
        options: []
        select: Label 2
      - key: field-3
        text: Label 3
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on 2`] = `
"steps:
  - key: _1
    block: my title
    fields:
      - key: release-name
        hint: 'What’s the code name for this release? :name_badge:'
        required: false
        text: Code Name
        default: Flying Dolphin
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on 3`] = `
"steps:
  - key: _2
    block: my title
    fields:
      - key: release-stream
        hint: What’s the release stream?
        required: false
        options:
          - label: Beta
            value: beta
          - label: Stable
            value: stable
        select: Stream
        default: beta
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on 4`] = `
"steps:
  - key: _3
    block: my title
    fields:
      - key: regions
        hint: 'Which regions should we deploy this to? :earth_asia:'
        options:
          - label: North America
            value: na
          - label: Europe
            value: eur
          - label: Asia
            value: asia
          - label: Oceania
            value: aunz
        select: Regions
        default:
          - na
          - eur
          - asia
          - aunz
        multiple: true
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on_accept_all 1`] = `
"steps:
  - key: _0
    block: my title
    fields:
      - key: field-1
        text: Label 1
      - key: field-2
        options: []
        select: Label 2
      - key: field-3
        text: Label 3
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on_accept_all 2`] = `
"steps:
  - key: _1
    block: my title
    fields:
      - key: release-name
        hint: 'What’s the code name for this release? :name_badge:'
        required: false
        text: Code Name
        default: Flying Dolphin
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on_accept_all 3`] = `
"steps:
  - key: _2
    block: my title
    fields:
      - key: release-stream
        hint: What’s the release stream?
        required: false
        options:
          - label: Beta
            value: beta
          - label: Stable
            value: stable
        select: Stream
        default: beta
"
`;

exports[`buildkite-graph Steps Block with fields yaml_depends_on_accept_all 4`] = `
"steps:
  - key: _3
    block: my title
    fields:
      - key: regions
        hint: 'Which regions should we deploy this to? :earth_asia:'
        options:
          - label: North America
            value: na
          - label: Europe
            value: eur
          - label: Asia
            value: asia
          - label: Oceania
            value: aunz
        select: Regions
        default:
          - na
          - eur
          - asia
          - aunz
        multiple: true
"
`;

exports[`buildkite-graph Steps Block with fields yaml_mutate 1`] = `
"steps:
  - key: _0
    block: my title
    fields:
      - key: field-1
        text: Label 1
      - key: field-2
        options: []
        select: Label 2
      - key: field-3
        text: Label 3
"
`;

exports[`buildkite-graph Steps Block with fields yaml_mutate 2`] = `
"steps:
  - key: _1
    block: my title
    fields:
      - key: release-name
        hint: 'What’s the code name for this release? :name_badge:'
        required: false
        text: Code Name
        default: Flying Dolphin
"
`;

exports[`buildkite-graph Steps Block with fields yaml_mutate 3`] = `
"steps:
  - key: _2
    block: my title
    fields:
      - key: release-stream
        hint: What’s the release stream?
        required: false
        options:
          - label: Beta
            value: beta
          - label: Stable
            value: stable
        select: Stream
        default: beta
"
`;

exports[`buildkite-graph Steps Block with fields yaml_mutate 4`] = `
"steps:
  - key: _3
    block: my title
    fields:
      - key: regions
        hint: 'Which regions should we deploy this to? :earth_asia:'
        options:
          - label: North America
            value: na
          - label: Europe
            value: eur
          - label: Asia
            value: asia
          - label: Oceania
            value: aunz
        select: Regions
        default:
          - na
          - eur
          - asia
          - aunz
        multiple: true
"
`;
